### PR TITLE
Remove bashfiles 

### LINKED ISSUE #91: feature(docker): Get rid of the mandatory bash file
See #86.

### PR DESCRIPTION
**Pull request by ash-sxn**
- Will partially fix #91 
- haven't tested it fully yet but spins up the correct containers with keywords `maven` `python` `node` with `docker compose up -d` 
- `docker compose up -d` for default (controller+agent)
- `docker compose up -d maven` for default (controller+maven agent)
- `docker compose up -d python ` for default (controller+python agent)
- `docker compose up -d node ` for default (controller+node agent)

### CODE REVIEW

File: dockerfiles/sidekick/keygen.sh [hunk 1]
```diff
@@ -0,0 +1,38 @@
+#!/bin/bash
+# This script generates a new SSH key pair and updates a Docker Compose file with the public key.
+# Create secrets directory if it doesn't exist
+set -x
+LOC=$1
+if [[ -z "$LOC" ]]; then
+    LOC="./"
+fi
+
+echo Location is $LOC
```
Comment (gounthar) [inline dockerfiles/sidekick/keygen.sh:10]:
```suggestion
echo "Location is $LOC"
```

File: dockerfiles/sidekick/keygen.sh [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+ls -artl $LOC
```
Comment (gounthar) [inline dockerfiles/sidekick/keygen.sh:11]:
```suggestion
ls -artl "$LOC"
```

File: dockerfiles/sidekick/keygen.sh [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+rm $LOC/conductor_ok
```
Comment (gounthar) [inline dockerfiles/sidekick/keygen.sh:13]:
```suggestion
rm "$LOC/conductor_ok"
```

File: dockerfiles/sidekick/keygen.sh [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+mkdir -p $LOC/secrets
```
Comment (gounthar) [inline dockerfiles/sidekick/keygen.sh:15]:
```suggestion
mkdir -p "$LOC/secrets"
```

File: dockerfiles/sidekick/keygen.sh [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+# Remove existing keys
+rm -fr $LOC/jenkins_agent_ed $LOC/jenkins_agent_ed.pub
```
Comment (gounthar) [inline dockerfiles/sidekick/keygen.sh:18]:
```suggestion
rm -fr "$LOC/jenkins_agent_ed" "$LOC/jenkins_agent_ed.pub"
```

File: dockerfiles/sidekick/keygen.sh [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+# Generate new ed25519 SSH key pair
+ssh-keygen -t ed25519 -f $LOC/jenkins_agent_ed -N ""
```
Comment (gounthar) [inline dockerfiles/sidekick/keygen.sh:21]:
```suggestion
ssh-keygen -t ed25519 -f "$LOC/jenkins_agent_ed" -N ""
```

File: dockerfiles/sidekick/keygen.sh [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+
+# Set appropriate permissions for private key
+chmod 444 $LOC/jenkins_agent_ed
```
Comment (gounthar) [inline dockerfiles/sidekick/keygen.sh:24]:
```suggestion
chmod 444 "$LOC/jenkins_agent_ed"
```

File: dockerfiles/sidekick/keygen.sh [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+
+# Extract public key
+pubkey=$(cat $LOC/jenkins_agent_ed.pub)
```
Comment (gounthar) [inline dockerfiles/sidekick/keygen.sh:27]:
```suggestion
pubkey=$(cat "$LOC/jenkins_agent_ed.pub")
```

File: dockerfiles/sidekick/keygen.sh [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+
+echo The public key is $pubkey
```
Comment (gounthar) [inline dockerfiles/sidekick/keygen.sh:29]:
```suggestion
echo "The public key is $pubkey"
```

File: dockerfiles/sidekick/keygen.sh [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+
+# Update the authorized_keys file with the public key
+echo "$pubkey" > $LOC/authorized_keys && chown 1000:1000 $LOC/authorized_keys
```
Comment (gounthar) [inline dockerfiles/sidekick/keygen.sh:32]:
```suggestion
echo "$pubkey" > "$LOC/authorized_keys" && chown 1000:1000 "$LOC/authorized_keys"
```

File: dockerfiles/sidekick/keygen.sh [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+
+# This file will be used by other containers to know we went up to the end of the key generation
+echo "OK" > $LOC/conductor_ok
```
Comment (gounthar) [inline dockerfiles/sidekick/keygen.sh:35]:
```suggestion
echo "OK" > "$LOC/conductor_ok"
```

**Review by gounthar — CHANGES_REQUESTED** [review]

**Comment by gounthar** [comment]
I have this error:

```
desktop-jenkins_agent-1: Name or service not known
SSH Connection failed with IOException: "desktop-jenkins_agent-1: Name or service not known", retrying in 15 seconds. There are 2 more retries left.
desktop-jenkins_agent-1: Name or service not known
SSH Connection failed with IOException: "desktop-jenkins_agent-1: Name or service not known", retrying in 15 seconds. There are 1 more retries left.
desktop-jenkins_agent-1: Name or service not known
ERROR: Connection is not established!
java.lang.IllegalStateException: Connection is not established!
	at com.trilead.ssh2.Connection.getRemainingAuthMethods(Connection.java:988)
	at com.cloudbees.jenkins.plugins.sshcredentials.impl.TrileadSSHPublicKeyAuthenticator.getRemainingAuthMethods(TrileadSSHPublicKeyAuthenticator.java:89)
	at com.cloudbees.jenkins.plugins.sshcredentials.impl.TrileadSSHPublicKeyAuthenticator.canAuthenticate(TrileadSSHPublicKeyAuthenticator.java:81)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:176)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.Spliterators$IteratorSpliterator.tryAdvance(Spliterators.java:1812)
	at java.base/java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:127)
	at java.base/java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:502)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:488)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.FindOps$FindOp.evaluateSequential(FindOps.java:150)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.findFirst(ReferencePipeline.java:543)
	at com.cloudbees.jenkins.plugins.sshcredentials.SSHAuthenticator.newInstance(SSHAuthenticator.java:222)
	at com.cloudbees.jenkins.plugins.sshcredentials.SSHAuthenticator.newInstance(SSHAuthenticator.java:173)
	at hudson.plugins.sshslaves.SSHLauncher.openConnection(SSHLauncher.java:878)
	at hudson.plugins.sshslaves.SSHLauncher.lambda$launch$0(SSHLauncher.java:434)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
```

I think the agent has not been launched.

**Comment by gounthar** [comment]
The GitHub Actions tests don't pass anymore (which was expected).

**Comment by ash-sxn** [comment]
The problem was with the name of the agent containers. I've fixed it now it's working for me.

**Comment by gounthar** [comment]
You could create another PR, and modify the linked issue (#91) to say it will be close when the next PR (enter the number) will be done.

**Comment by gounthar** [comment]
And maybe another PR for the tests changes...

**Comment by gounthar** [comment]
Maven worked for me on GitPod.
Let me try it under Windows.

**Comment by gounthar** [comment]
It failed:

```
docker compose up maven
[+] Building 53.6s (10/10) FINISHED
 => [internal] load build definition from Dockerfile                                                                                                                          0.2s
 => => transferring dockerfile: 482B                                                                                                                                          0.0s
 => [internal] load .dockerignore                                                                                                                                             0.1s
 => => transferring context: 2B                                                                                                                                               0.0s
 => [internal] load metadata for docker.io/library/debian:bookworm                                                                                                            2.9s
 => [auth] library/debian:pull token for registry-1.docker.io                                                                                                                 0.0s
 => [internal] load build context                                                                                                                                             0.2s
 => => transferring context: 1.07kB                                                                                                                                           0.0s
 => [1/5] FROM docker.io/library/debian:bookworm@sha256:3d868b5eb908155f3784317b3dda2941df87bbbbaa4608f84881de66d9bb297b                                                     17.8s
 => => resolve docker.io/library/debian:bookworm@sha256:3d868b5eb908155f3784317b3dda2941df87bbbbaa4608f84881de66d9bb297b                                                      0.1s
 => => sha256:d52e4f012db158bb7c0fe215b98af1facaddcbaee530efd69b1bae07d597b711 49.55MB / 49.55MB                                                                             13.7s
 => => extracting sha256:d52e4f012db158bb7c0fe215b98af1facaddcbaee530efd69b1bae07d597b711                                                                                     3.9s
 => [2/5] COPY *sh /usr/local/bin/                                                                                                                                            0.4s
 => [3/5] RUN chmod +x /usr/local/bin/*.sh                                                                                                                                    0.3s
 => [4/5] RUN apt update && apt install -y --no-install-recommends     ca-certificates     curl     git     gnupg     nano     openssh-client     procps     unzip     wget  31.7s
 => ERROR [5/5] RUN /usr/local/bin/keygen.sh /ssh-dir                                                                                                                         0.2s
------
 > [5/5] RUN /usr/local/bin/keygen.sh /ssh-dir:
#0 0.173 /bin/sh: 1: /usr/local/bin/keygen.sh: not found
------
failed to solve: process "/bin/sh -c /usr/local/bin/keygen.sh /ssh-dir" did not complete successfully: exit code: 127
```

**Comment by gounthar** [comment]
I have the same kind of problems within WLS2.
I'll try that on a real Linux machine.

**Comment by gounthar** [comment]
It now works on my Windows machine, after cleaning up all my docker images, volumes, and containers with the `docker compose up --build --force-recreate maven` command... after the infamous dos2unix command on `keygen.sh`.
I don't know how to solve this one.

**Review by gounthar — APPROVED** [review]

**Comment by gounthar** [comment]
It also works on my `aarch64` Linux machine.

**Review by berviantoleo — APPROVED** [review]
Wow. You are guys fast at work.

So far looks good.

🎉 

Thanks @ash-sxn & @gounthar 

---

I think we will need to fix the GitHub Action after this.

**Comment by ash-sxn** [comment]
I was in docs office hours meeting with Mark & Meg just now. and we were trying this file on Windows and we got the same error as @gounthar 
```
=> ERROR [5/5] RUN /usr/local/bin/keygen.sh /ssh-dir                                                                                                                         0.2s
------
 > [5/5] RUN /usr/local/bin/keygen.sh /ssh-dir:
#0 0.173 /bin/sh: 1: /usr/local/bin/keygen.sh: not found
------
failed to solve: process "/bin/sh -c /usr/local/bin/keygen.sh /ssh-dir" did not complete successfully: exit code: 127
```
Mark recommended a quick fix by adding a .gitattributes file with this command `*.sh text elo=lf` and it worked 🎉
We tested two of the tutorial i.e node(windows with newly installed docker desktop) and python (on git bash) 

Thankyou @MarkEWaite 🙏🏼

**Comment by gounthar** [comment]
That's great news, thanks a lot ,!

**Pull request merged**